pkg/networkmanager/interfaces.js

```diff
@@ -612,31 +628,61 @@ export function NetworkManagerModel() {
                 delete result[first][second];
         }
 
-        function set_ip(first, addrs_sig, addr_to_nm, routes_sig, route_to_nm, ips_sig, ip_from_text) {
+        function set_ip(first, dns_ip_sig, ip_from_text) {
             set(first, "method", 's', settings[first].method);
             set(first, "ignore-auto-dns", 'b', settings[first].ignore_auto_dns);
             set(first, "ignore-auto-routes", 'b', settings[first].ignore_auto_routes);
             set(first, "addr-gen-mode", 'i', settings[first].addr_gen_mode);
 
-            const addresses = settings[first].addresses;
+            const addresses = settings[first].address_data;
             if (addresses)
-                set(first, "addresses", addrs_sig, addresses.map(addr_to_nm));
+                set(first, "address-data", "aa{sv}", addresses.map(addr => ip_address_to_nm(addr, first)));
+
+            const gateway = settings[first].gateway;
+            if (gateway && addresses.length > 0) {
+                if (!utils.validate_ip(gateway)) {
+                    throw cockpit.format(_("Invalid gateway address: $0"), gateway);
+                }
+                set(first, "gateway", "s", gateway);
+            } else {
+                // gateway cannot be set if there are no addresses
+                delete result[first].gateway;
+            }
+
+            const dns = settings[first].dns_data;
+            if (dns) {
+                const invalid = dns.find(addr => !utils.validate_ip(addr));
+                if (invalid) {
+                    throw cockpit.format(_("Invalid DNS address: $0"), invalid);
+                }
+
+                if (self.supports_dns_data) {
+                    set(first, "dns-data", "as", dns);
+                } else {
+                    set(first, "dns", dns_ip_sig, dns.map(addr => ip_from_text(addr)));
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22330-e10cfba6-20250915-113603-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)